### PR TITLE
[Feature][Diffusion] Enable BOOGU Image SP2 with 1.67x E2E Speedup

### PR DESCRIPTION
**Pull request by Xenoryn**
## Summary

Enable Ulysses sequence parallelism (SP2) for BOOGU Image.

BOOGU Base uses 28 query heads and 7 KV heads, so SP2 requires GQA-aware head padding through `advanced_uaa`. This PR also removes a per-attention host synchronization when sequence shards are already equally padded by the framework SP hooks.

### Sequence-parallel framework

- Add keyed shard metadata for models containing multiple independent sequences.
- Track equal-rank sequence lengths across framework-managed `auto_pad` split/gather regions.
- Skip UAA's runtime sequence-length all-gather while this equal-rank contract is active.
- Preserve the dynamic-length fallback for manually sharded or uneven inputs.
- Pad BOOGU's 28 Q / 7 KV heads to 32 / 8 at SP2, preserving its 4:1 GQA ratio.
- Support both global and rank-local 2D attention masks.
- Validate invalid or ambiguous shard-group reuse.

### BOOGU Image

- Define `image`, `reference`, and `context` shard groups through `_sp_plan`.
- Repack rank-local masks and RoPE tensors using framework shard metadata.
- Preserve padded capacities through the refiner and joint-stream stages.
- Gather only the final noise-image tokens and trim them to the original length.

The BOOGU adapter does not introduce process-group calls or communication-backend branches.

## Limitations

- BOOGU currently supports pure Ulysses SP only.
- SP requires CUDA/NCCL.
- Ring, hybrid USP, and AllGather-KV are not supported by this PR.

### Tests

```bash
pytest -q \
  tests/diffusion/distributed/test_sp_plan_hooks.py \
  tests/diffusion/models/boogu_image/test_pipeline_boogu_image.py \
  tests/diffusion/attention/test_ulysses_uaa.py

pytest -q -s \
  tests/diffusion/distributed/test_boogu_image_sp.py
```

Coverage includes:

- keyed metadata and independent auto-padding;
- equal-rank fast path and unequal-length fallback;
- GQA head padding;
- global and rank-local masks;
- BOOGU SP1/SP2 correctness for text-to-image and image editing.

Results:

- SP hooks, validation, and BOOGU pipeline: **106 passed**
- UAA regression: **6 passed, 3 skipped**
- Tiny BOOGU SP1/SP2 comparison: **passed**
- Global/local mask comparisons: **passed**
- Ruff, Python compilation, marker validation, and `git diff --check`: **passed**

## Performance

Tested on 2 × NVIDIA H800 with BF16, FlashAttention, `advanced_uaa`, and `torch.compile(dynamic=True)`.

### Compiled BOOGU block

Both configurations use the same SP2 execution path. The only difference is how UAA obtains rank-local sequence lengths:

- **SP2 with length sync:** performs a host-side length all-gather in every attention call.
- **SP2 with known equal lengths:** reuses the equal-length guarantee established by framework-managed padding and avoids that synchronization.

| Resolution | Batch | SP2 with length sync | SP2 with known equal lengths | Speedup |
|---:|---:|---:|---:|---:|
| 1024² | 1 | 2.822 ms | 2.024 ms | +39.43% |
| 1024² | 2 | 4.798 ms | 3.760 ms | +27.59% |
| 1024² | 4 | 9.089 ms | 7.475 ms | +21.59% |
| 1024² | 8 | 17.226 ms | 14.318 ms | +20.32% |
| 1536² | 1 | 5.361 ms | 4.600 ms | +16.55% |
| 1536² | 2 | 10.843 ms | 9.050 ms | +19.82% |
| 1536² | 4 | 20.901 ms | 17.637 ms | +18.51% |
| 2048² | 1 | 10.395 ms | 9.356 ms | +11.11% |
| 2048² | 2 | 21.039 ms | 18.196 ms | +15.63% |

Removing the redundant length synchronization produces a **20.94% geometric-mean block speedup**.

### End-to-end: 1024², 50 steps

| Configuration | Latency | Improvement |
|---|---:|---:|
| Main SP1 | 15.138 s | baseline |
| This PR: SP2 | **9.033 s** | **40.33% lower latency** |

SP2 is **1.68× faster** than main SP1. Repeated runs produced identical RGB pixel hashes within each configuration.

<details>
<summary>Visual comparison and benchmark environment</summary>

The outputs were generated with the same checkpoint, prompt, guidance scale, step count, and seed. Minor BF16 distributed-order differences are expected; visual inspection confirmed that composition and prompt semantics were preserved.

| SP1 | SP2 |
|---|---|
| BOOGU SP1 output | BOOGU SP2 output |

- Model: `/workspace/local-models/Boogu-Image-0.1-Base`
- vLLM: `0.24.0`
- vLLM-Omni base commit: `e043818abfc060903a188bde01b9ce648e634054`
- Python: `3.12.3`
- PyTorch: `2.11.0+cu130`
- CUDA: `13.0`
- NVIDIA driver: `570.148.08`
- Hardware: 2 × NVIDIA H800
- E2E configuration: batch size 1, guidance scale 4.0, seed 42
- Timing: one 4-step warmup followed by three 50-step requests

</details>

### CODE REVIEW

**Comment by hsliuustc0106** [comment]
can you provide the comm proportion with SP2 or SP4?

**Comment by Xenoryn** [comment]
> can you provide the comm proportion with SP2 or SP4?

@hsliuustc0106 Thanks — sorry for the delay. This draft has since been split into three individually-reviewable PRs (#5716, #5717, #5718); the comm analysis you asked for lives on the end-user PR:

https://github.com/vllm-project/vllm-omni/pull/5718#issuecomment-5538475652

TL;DR: SendRecv (Ulysses all-to-all) is the only meaningful comm kernel; SP2 ≈ 11–19 % of GPU time, SP4 Base ≈ 60 %, SP4 Edit ≈ 32 %;
